File: 0200-number-of-islands/0200-number-of-islands.py
def visit(grid, i, j):
    if grid[i][j] == '0':
        return

    # (in-place) visited를 쓰지 않고 기존 값을 0으로 바꾼다
    grid[i][j] = "0"

    # 상하좌우 움직임을 direction으로 정의해서 순회한다.
    for di, dj in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        ni, nj = i + di, j + dj
        if 0 <= ni < len(grid) and 0 <= nj < len(grid[0]):
            visit(grid, ni, nj)

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ans = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    visit(grid, i, j)
                    ans += 1

        return ans

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    static int[][] directions = {{1,0}, {0,1}, {-1,0}, {0,-1}};

    public int numIslands(char[][] grid) {
        int n = grid.length, m = grid[0].length;
        boolean[][] visited = new boolean[n][m];

        int count = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    visit(grid, visited, i, j, n, m);
                    count++;
                }
            }
        }

        return count;
    }

    private void visit(char[][] grid, boolean[][] visited, int x, int y, int n, int m) {
        visited[x][y] = true;

        for (int[] d: directions) {
            int nx = d[0] + x, ny = d[1] + y;
            if (
                (0 <= nx && nx < n) && (0 <= ny && ny < m) 
                && !visited[nx][ny] && grid[nx][ny] == '1'
            ) {
                visit(grid, visited, nx, ny, n, m);
            }
        }
    }
}
